### PR TITLE
*: fix asyncGetTSWorker goroutine leak

### PR DESCRIPTION
**Pull request by tiancaiamao**
sysSessionPool use sync.Pool to reuse session object.
howerer, objects in the sync.Pool may be free by GC, its Close() won't be called.
so the asyncGetTSWorker goroutine bind with the session leak.

replace sync.Pool with another implementation to fix.

### CODE REVIEW

File: tidb_test.go [hunk 1]
```diff
@@ -344,6 +345,30 @@ func (s *testMainSuite) TestSchemaValidity(c *C) {
...
+		go func(se *session) {
+			_, _, err := se.ExecRestrictedSQL(se, "select * from mysql.user limit 1")
+			c.Assert(err, IsNil)
+			wg.Done()
+		}(se)
+	}
+	wg.Wait()
+	se.sysSessionPool().Close()
+	after := runtime.NumGoroutine()
+	c.Assert(after-before < 3, IsTrue)
```
Comment (shenli) [inline tidb_test.go]: Why check this?

File: _vendor/src/github.com/ngaut/pools/resource_pool.go [hunk 2]
```diff
@@ -24,7 +24,7 @@ type Factory func() (Resource, error)
 // Thread synchronization between Close() and IsClosed()
 // is the responsibility the caller.
 type Resource interface {
-	Close()
+	Close() error
```
Comment (shenli) [inline _vendor/src/github.com/ngaut/pools/resource_pool.go]: It is not necessary to return error.

File: domain/domain.go [hunk 3]
```diff
@@ -356,12 +358,12 @@ type etcdBackend interface {
...
 
 // NewDomain creates a new domain. Should not create multiple domains for the same store.
-func NewDomain(store kv.Storage, lease time.Duration) (d *Domain, err error) {
+func NewDomain(store kv.Storage, lease time.Duration, factory pools.Factory) (d *Domain, err error) {
 	d = &Domain{
 		store:           store,
 		SchemaValidator: newSchemaValidator(lease),
 		exit:            make(chan struct{}),
-		sysSessionPool:  &sync.Pool{},
+		sysSessionPool:  pools.NewResourcePool(factory, 1, 200, 3*time.Minute),
```
Comment (shenli) [inline domain/domain.go]: Can we create the factory inside NewDomain?

File: tidb_test.go (hunk 1 above)
Comment (tiancaiamao) [inline tidb_test.go]: check no goroutine leak

File: domain/domain.go (hunk 3 above)
Comment (tiancaiamao) [inline domain/domain.go]: domain package don't have access to *session, so the factory can't be defined inside this package.

**Comment by tiancaiamao** [comment]
PTAL @shenli @zimulala @hanfei1991 @XuHuaiyu

File: tidb_test.go (hunk 1 above)
Comment (siddontang) [inline tidb_test.go]:
```
c.Assert(after- before, Less, 3)
```

File: session.go [hunk 4]
```diff
@@ -783,14 +788,17 @@ func (s *session) ClearValue(key fmt.Stringer) {
 }
 
 // Close function does some clean work when session end.
-func (s *session) Close() error {
+func (s *session) Close() {
```
Comment (zimulala) [inline session.go]: Why change this interface?

Comment (tiancaiamao) [inline session.go]: Because "github.com/ngaut/pools" define the Resource interface to be 

```
type Resource interface {
    Close()
}
```

Comment (zimulala) [inline session.go]: I got it.

**Comment by zimulala** [comment]
LGTM

**Comment by tiancaiamao** [comment]
PTAL @shenli

File: domain/domain.go [hunk 5]
```diff
@@ -356,12 +358,15 @@ type etcdBackend interface {
 }
 
 // NewDomain creates a new domain. Should not create multiple domains for the same store.
-func NewDomain(store kv.Storage, lease time.Duration) (d *Domain, err error) {
+func NewDomain(store kv.Storage, lease time.Duration, factory pools.Factory) (d *Domain, err error) {
+	minCapacity := 1               // minCapacity for the sysSessionPool size
+	maxCapacity := 200             // maxCapacity for the sysSessionPool size
```
Comment (shenli) [inline domain/domain.go]: Should we enlarge maxCapacity to 500 or 1000? 
https://github.com/pingcap/tidb/blob/master/server/server.go#L126

Comment (tiancaiamao) [inline domain/domain.go]: OK, I'll enlarge it to 500.
ExecuteRestrictedSQL use session from this pool, it should not be too much, I think.

File: tidb_test.go [hunk 6]
```diff
@@ -332,18 +333,39 @@ func (s *testMainSuite) TestSchemaValidity(c *C) {
...
+		go func(se *session) {
+			_, _, err := se.ExecRestrictedSQL(se, "select * from mysql.user limit 1")
+			c.Assert(err, IsNil)
+			wg.Done()
+		}(se)
+	}
+	wg.Wait()
+	se.sysSessionPool().Close()
+	after := runtime.NumGoroutine()
+	c.Assert(after-before, Less, 3)
```
Comment (shenli) [inline tidb_test.go]: Why not after equals before?

Comment (tiancaiamao) [inline tidb_test.go]: It should be Equal, but I'm afraid this test disturbed by other factors that would make CI unstable, so I relax the strict check.

**Comment by shenli** [comment]
LGTM

**Comment by shenli** [comment]
@tiancaiamao Please fix the conflicts.

**Pull request merged**